### PR TITLE
Simplify torch.nn.grad by calling into aten::convolution_backward

### LINKED ISSUE #51430: Convolution weight gradient computation in nn.grad is wrong with groups > 1
## 🐛 Bug

Convolution weight gradient computation in nn.grad seems to be wrong when groups > 1.
(conv1d_weight, conv2d_weight, conv3d_weight) 

## To Reproduce

Steps to reproduce the behavior: I attached my gist to print the difference between autograd and nn.grad.
In the script, I also suggest a correct alternative (only a minor change was sufficient)

https://gist.github.com/lemonbuffer20/7c98cea1e904b2abaa38ce6d4ec9605e

## Expected behavior

nn.grad.convNd_weight should return the same result as F.conv2d backprop in groups > 1 case.

## Environment

PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX TITAN X

Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8
...

## Additional context

1.
I believe the small difference (~=1e-6) of the result is inevitable.
But sometimes the max grad difference (both input and weight) goes to ~=1e-4, which seems much bigger.
Can we remove, or at least reduce the error?

2.
If you can verify my modification is correct, I'd be glad to propose PR.

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @kshitij12345 @bdhirsh

### PR DESCRIPTION
**Pull request by saketh-are**
`torch.nn.grad` has its own implementations of gradients for conv1d, conv2d, and conv3d. This PR simplifies them by calling into the unified `aten::convolution_backward` backend instead.

The existing implementation of conv2d_weight is incorrect for some inputs (see issue #51430). This PR fixes the issue.

This PR expands coverage in test_nn to include conv1d_weight, conv2d_weight, and conv3d_weight, which were previously untested. It also expands the cases for conv2d to cover issue #51430.

Fixes #51430

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -11771,34 +11771,85 @@ def test_functional_grad_conv(self):
         output = F.conv1d(input, weight, dilation=2)
         grad_output = torch.randn(output.shape)
 
-        grad_input_autograd = torch.autograd.grad(output, input, grad_output)[0]
+        grad_autograd = torch.autograd.grad(output, (input, weight), grad_output)
```
Comment (albanD) [inline test/test_nn.py:11774]: A nice solution for this is to do `grad_inp, grad_weight = autograd.grad(...)`

File: torch/nn/grad.py [hunk 2]
```diff
@@ -101,29 +59,11 @@ def conv1d_weight(input, weight_size, grad_output, stride=1, padding=0, dilation
...
-    out_channels = grad_output.shape[1]
-    min_batch = input.shape[0]
-
-    grad_output = grad_output.contiguous().repeat(1, in_channels // groups, 1)
-    grad_output = grad_output.contiguous().view(
-        grad_output.shape[0] * grad_output.shape[1], 1, grad_output.shape[2])
-
-    input = input.contiguous().view(1, input.shape[0] * input.shape[1],
-                                    input.shape[2])
+    weight = grad_output.new_empty(weight_size)
```
Comment (albanD) [inline torch/nn/grad.py:62]: here and below you can make this more efficient by not allocating a full Tensor:
```python
weight = grad_putput.new_empty(1).expand(weight_size)

```

File: test/test_nn.py [hunk 3]
```diff
@@ -11771,34 +11771,85 @@ def test_functional_grad_conv(self):
...
+            grad_weight_functional = torch.nn.grad.conv2d_weight(input, weight.shape, grad_output,
+                                                                 stride=stride, padding=padding, dilation=dilation, groups=groups)
+            self.assertEqual(grad_weight_functional, grad_weight_autograd)
+
+        strides = [1, 2]
+        kernel_sizes = [1, 3, 5]
+        groups = [1, 2, 4]
+        dilates = [1, 2]
+
+        for s, k, g, d in product(strides, kernel_sizes, groups, dilates):
```
Comment (albanD) [inline test/test_nn.py:11851]: That is a lot of tests? How long does this test takes to run?

**Review by albanD — COMMENTED** [review]
Sounds ok.

File: test/test_nn.py (hunk 3 above)
Comment (saketh-are) [reply test/test_nn.py:11851]: 0.237s total on my local machine

File: torch/nn/grad.py (hunk 2 above)
Comment (saketh-are) [reply torch/nn/grad.py:62]: Thanks, this is because we will only be reading from `weight` tensor and `expand` creates a view with the right shape instead of actually allocating it?

File: test/test_nn.py (hunk 3 above)
Comment (albanD) [reply test/test_nn.py:11843]: Less than a second sounds ok for this.

File: torch/nn/grad.py (hunk 2 above)
Comment (albanD) [reply torch/nn/grad.py]: Exactly, since we never use the content anyways, we can make it a single random value instead of allocating a full size Tensor of random values.

File: test/test_nn.py [hunk 4]
```diff
@@ -11771,34 +11771,77 @@ def test_functional_grad_conv(self):
         output = F.conv1d(input, weight, dilation=2)
         grad_output = torch.randn(output.shape)
 
-        grad_input_autograd = torch.autograd.grad(output, input, grad_output)[0]
+        (grad_input_autograd, grad_weight_autograd) = torch.autograd.grad(output, (input, weight), grad_output)
```
Comment (albanD) [inline test/test_nn.py:11774]: nit: the parenthesis on the left side are not needed.

**Review by albanD — APPROVED** [review]
SGTM!
